bumble/pandora/l2cap.py

```diff
@@ -0,0 +1,318 @@
+            assert request.le_credit_based
+            channel = await connection.create_l2cap_channel(
+                spec=l2cap.LeCreditBasedChannelSpec(
+                    psm=request.le_credit_based.spsm,
+                    max_credits=request.le_credit_based.initial_credit,
+                    mtu=request.le_credit_based.mtu,
+                    mps=request.le_credit_based.mps,
+                )
+            )
+        else:
+            raise NotImplementedError
+
+        self.channels[connection_handle][channel.source_cid] = CocChannelProxy(channel)
+        l2cap_cookie = any_pb2.Any(
+            value=struct.pack('>HH', connection_handle, channel.source_cid)
+        )
+        return l2cap_pb2.ConnectResponse(channel=l2cap_pb2.Channel(cookie=l2cap_cookie))
+
+    @utils.rpc
+    async def OnConnection(
+        self, request: l2cap_pb2.OnConnectionRequest, context: grpc.ServicerContext
+    ) -> AsyncGenerator[l2cap_pb2.OnConnectionResponse, None]:
+        self.log.info('WaitConnection')
+
+        queue: asyncio.Queue[l2cap_pb2.OnConnectionResponse] = asyncio.Queue()
+        connection_handle = int.from_bytes(request.connection.cookie.value, 'big')
+
+        connection = self.device.lookup_connection(connection_handle)
+        if connection is None:
+            raise RuntimeError('Connection not exist')
+
+        self.channels.setdefault(connection_handle, {})
+
+        watcher = EventWatcher()
+        server: Union[
+            l2cap.ClassicChannelServer, l2cap.LeCreditBasedChannelServer, None
+        ] = None
+        fixed_cid: Optional[int] = None
+
+        # Fixed channels are connectionless, so it should produce a response immediately.
+        if request.type_variant() == 'fixed':
+            assert request.fixed
+            cid = request.fixed.cid
+
+            def on_fixed_pdu(connection_handle: int, pdu: bytes) -> None:
+                self.channels[connection_handle][cid].on_data(pdu)
+
+            channel_proxy = FixedChannelProxy(
+                connection_handle=connection_handle, cid=cid, device=self.device
+            )
+            self.channels[connection_handle][cid] = channel_proxy
+            l2cap_cookie = any_pb2.Any(value=struct.pack('>HH', connection_handle, cid))
+
+            # Register CID and callback
+            self.device.l2cap_channel_manager.register_fixed_channel(cid, on_fixed_pdu)
+
+            queue.put_nowait(
+                l2cap_pb2.OnConnectionResponse(
+                    channel=l2cap_pb2.Channel(cookie=l2cap_cookie)
+                )
+            )
+        else:
+
+            def on_connected(
+                channel: Union[l2cap.ClassicChannel, l2cap.LeCreditBasedChannel]
+            ) -> None:
+                handle = channel.connection.handle
+                # Filter channels from non-specified connection
+                if handle != connection_handle:
+                    connection.abort_on('disconnect', channel.disconnect())
```

Comment:
It should be possible to have multiple `OnConnection` in parallel, we should not disconnect when filtering

Reply:
Well, that's the problem of this interface. Since Bumble L2CAP server is based on CID instead of connection-cid pair, there must be no more than 1 server for the same cid. The same restriction also exists on Android.